File: sdk/python/polar/payments.py
from __future__ import annotations

import builtins
import typing

from polar.base import AsyncServiceBase, SyncServiceBase, parse_response_json
from polar.errors import (
    HTTPValidationError,
    ResourceNotFound,
)
from polar.literals import (
    PaymentSortProperty,
    PaymentStatus,
)
from polar.outputs import (
    ListResourcePayment,
    Payment,
)


class PaymentsSync(SyncServiceBase):
    def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        checkout_id: str | builtins.list[str] | None = None,
        order_id: str | builtins.list[str] | None = None,
        customer_id: str | builtins.list[str] | None = None,
        status: PaymentStatus | builtins.list[PaymentStatus] | None = None,
        method: str | builtins.list[str] | None = None,
        customer_email: str | builtins.list[str] | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[PaymentSortProperty] | None = ["-created_at"],
    ) -> ListResourcePayment:
        """
        List payments.

        **Scopes**: `payments:read`

        Args:
            organization_id: Filter by organization ID.
            checkout_id: Filter by checkout ID.
            order_id: Filter by order ID.
            customer_id: Filter by customer ID.
            status: Filter by payment status.
            method: Filter by payment method.
            customer_email: Filter by customer email.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/payments/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "checkout_id": checkout_id,
                "order_id": order_id,
                "customer_id": customer_id,
                "status": status,
                "method": method,
                "customer_email": customer_email,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, ListResourcePayment, method_errors)

    def iter_list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        checkout_id: str | builtins.list[str] | None = None,
        order_id: str | builtins.list[str] | None = None,
        customer_id: str | builtins.list[str] | None = None,
        status: PaymentStatus | builtins.list[PaymentStatus] | None = None,
        method: str | builtins.list[str] | None = None,
        customer_email: str | builtins.list[str] | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[PaymentSortProperty] | None = ["-created_at"],
    ) -> typing.Generator[Payment]:
        """
        List payments.

        **Scopes**: `payments:read`

        Args:
            organization_id: Filter by organization ID.
            checkout_id: Filter by checkout ID.
            order_id: Filter by order ID.
            customer_id: Filter by customer ID.
            status: Filter by payment status.
            method: Filter by payment method.
            customer_email: Filter by customer email.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Returns:
            A generator that yields items of type Payment.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        while True:
            response = self.list(
                organization_id=organization_id,
                checkout_id=checkout_id,
                order_id=order_id,
                customer_id=customer_id,
                status=status,
                method=method,
                customer_email=customer_email,
                page=page,
                limit=limit,
                sorting=sorting,
            )
            yield from response.items
            if page == response.pagination.max_page:
                break
            page += 1

    def get(
        self,
        id: str,
    ) -> Payment:
        """
        Get a payment by ID.

        **Scopes**: `payments:read`

        Args:
            id: The payment ID.

        Raises:
            ResourceNotFound: Payment not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/payments/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_json(response, Payment, method_errors)


class PaymentsAsync(AsyncServiceBase):
    async def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        checkout_id: str | builtins.list[str] | None = None,
        order_id: str | builtins.list[str] | None = None,
        customer_id: str | builtins.list[str] | None = None,
        status: PaymentStatus | builtins.list[PaymentStatus] | None = None,
        method: str | builtins.list[str] | None = None,
        customer_email: str | builtins.list[str] | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[PaymentSortProperty] | None = ["-created_at"],
    ) -> ListResourcePayment:
        """
        List payments.

        **Scopes**: `payments:read`

        Args:
            organization_id: Filter by organization ID.
            checkout_id: Filter by checkout ID.
            order_id: Filter by order ID.
            customer_id: Filter by customer ID.
            status: Filter by payment status.
            method: Filter by payment method.
            customer_email: Filter by customer email.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/payments/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "checkout_id": checkout_id,
                "order_id": order_id,
                "customer_id": customer_id,
                "status": status,
                "method": method,
                "customer_email": customer_email,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, ListResourcePayment, method_errors)

    async def iter_list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        checkout_id: str | builtins.list[str] | None = None,
        order_id: str | builtins.list[str] | None = None,
        customer_id: str | builtins.list[str] | None = None,
        status: PaymentStatus | builtins.list[PaymentStatus] | None = None,
        method: str | builtins.list[str] | None = None,
        customer_email: str | builtins.list[str] | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[PaymentSortProperty] | None = ["-created_at"],
    ) -> typing.AsyncGenerator[Payment]:
        """
        List payments.

        **Scopes**: `payments:read`

        Args:
            organization_id: Filter by organization ID.
            checkout_id: Filter by checkout ID.
            order_id: Filter by order ID.
            customer_id: Filter by customer ID.
            status: Filter by payment status.
            method: Filter by payment method.
            customer_email: Filter by customer email.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Returns:
            An async generator that yields items of type Payment.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        while True:
            response = await self.list(
                organization_id=organization_id,
                checkout_id=checkout_id,
                order_id=order_id,
                customer_id=customer_id,
                status=status,
                method=method,
                customer_email=customer_email,
                page=page,
                limit=limit,
                sorting=sorting,
            )
            for item in response.items:
                yield item
            if page == response.pagination.max_page:
                break
            page += 1

    async def get(
        self,
        id: str,
    ) -> Payment:
        """
        Get a payment by ID.

        **Scopes**: `payments:read`

        Args:
            id: The payment ID.

        Raises:
            ResourceNotFound: Payment not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/payments/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_json(response, Payment, method_errors)

File: sdk/typescript/src/2026-04/services/payments.ts
import type { ClientBase } from "../../base";
import type { ListResourcePayment, Payment, PaymentSortProperty, PaymentStatus } from "../models";

import { HTTPValidationError, ResourceNotFound } from "../errors";

export const listPayments = (client: ClientBase) => {
  /**
   * List payments.
   *
   * **Scopes**: `payments:read`
   *
   * @param query - Query parameters
   * @returns {ListResourcePayment}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query?: {
    organization_id?: string | string[] | null;
    checkout_id?: string | string[] | null;
    order_id?: string | string[] | null;
    customer_id?: string | string[] | null;
    status?: PaymentStatus | PaymentStatus[] | null;
    method?: string | string[] | null;
    customer_email?: string | string[] | null;
    page?: number;
    limit?: number;
    sorting?: PaymentSortProperty[] | null;
  }): Promise<ListResourcePayment> => {
    const pathParams = {};
    const queryParams = {
      organization_id: query?.organization_id,
      checkout_id: query?.checkout_id,
      order_id: query?.order_id,
      customer_id: query?.customer_id,
      status: query?.status,
      method: query?.method,
      customer_email: query?.customer_email,
      page: query?.page ?? 1,
      limit: query?.limit ?? 10,
      sorting: query?.sorting ?? ["-created_at"],
    };
    const request = client.buildRequest("GET", "/v1/payments/", pathParams, queryParams, undefined);
    const response = await client.sendRequest(request);
    return client.parseResponse<ListResourcePayment>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
/**
 * List payments.
 *
 * **Scopes**: `payments:read`
 *
 * @param query - Query parameters
 * @returns {AsyncGenerator<Payment>} A generator that yields items of type Payment.
 * @throws {PolarNetworkError} When a network error occurs
 * @throws {PolarRateLimitError} When the rate limit is exceeded
 * @throws {PolarServerError} When the server returns a 5xx error
 * @throws {HTTPValidationError} Validation Error
 */
export const iterlistPayments = (client: ClientBase) => {
  return async function* (query?: {
    organization_id?: string | string[] | null;
    checkout_id?: string | string[] | null;
    order_id?: string | string[] | null;
    customer_id?: string | string[] | null;
    status?: PaymentStatus | PaymentStatus[] | null;
    method?: string | string[] | null;
    customer_email?: string | string[] | null;
    page?: number;
    limit?: number;
    sorting?: PaymentSortProperty[] | null;
  }): AsyncGenerator<Payment> {
    let page: number;
    page = query?.page ?? 1;
    let limit: number | undefined;
    limit = query?.limit;

    while (true) {
      const response = await listPayments(client)({ ...query, page, limit });
      for (const item of response.items) {
        yield item;
      }
      if (page >= response.pagination.max_page) {
        break;
      }
      page++;
    }
  };
};
export const getPayments = (client: ClientBase) => {
  /**
   * Get a payment by ID.
   *
   * **Scopes**: `payments:read`
   *
   * @param id - The payment ID.
   * @returns {Payment}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Payment not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string): Promise<Payment> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/payments/{id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<Payment>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};

export function createPaymentsService(client: ClientBase) {
  return {
    list: listPayments(client),
    get: getPayments(client),
    iterlist: iterlistPayments(client),
  };
}

export type Payments = ReturnType<typeof createPaymentsService>;
